SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestNetReceive/SpatialTestNetReceive.h

```diff
@@ -0,0 +1,82 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#pragma once
+
+#include "CoreMinimal.h"
+#include "SpatialFunctionalTest.h"
+#include "SpatialGDKFunctionalTests/SpatialGDK/TestActors/ReplicatedTestActorBase.h"
+#include "SpatialTestNetReceive.generated.h"
+
+class ASpatialTestNetReceiveActor;
+class USpatialTestNetReceiveSubobject;
+
+UCLASS()
+class SPATIALGDKFUNCTIONALTESTS_API ASpatialTestNetReceive : public ASpatialFunctionalTest
+{
+	GENERATED_BODY()
+
+public:
+	ASpatialTestNetReceive();
+
+	void GetLifetimeReplicatedProps(TArray<FLifetimeProperty>& OutLifetimeProps) const override;
+
+	void PrepareTest() override;
+
+	UPROPERTY(Replicated)
+	ASpatialTestNetReceiveActor* TestActor;
+};
+
+UCLASS()
+class ASpatialTestNetReceiveActor : public AReplicatedTestActorBase
+{
+	GENERATED_BODY()
+
+public:
+	ASpatialTestNetReceiveActor();
+
+	void GetLifetimeReplicatedProps(TArray<FLifetimeProperty>& OutLifetimeProps) const override;
+
+	UPROPERTY(Replicated)
+	USpatialTestNetReceiveSubobject* Subobject;
+};
+
+enum class RepStep
+{
+	PreRep,
+	PreNetReceive,
+	PostNetReceive,
+	RepNotify,
+
+	SecondPreNetReceive, // For this test, expected to have RepNotify called before it
+};
+
+UCLASS()
+class SPATIALGDKFUNCTIONALTESTS_API USpatialTestNetReceiveSubobject : public UActorComponent
+{
+	GENERATED_BODY()
+
+public:
+	USpatialTestNetReceiveSubobject();
+
+	void GetLifetimeReplicatedProps(TArray<FLifetimeProperty>& OutLifetimeProps) const override;
+
+	void PreNetReceive() override;
+	void PostNetReceive() override;
+
+	UPROPERTY(ReplicatedUsing = OnRep_TestInt)
+	int32 TestInt;
+
+	// Use an owner only test int to ensure two spatial components are added and updated
+	UPROPERTY(Replicated)
+	int32 OwnerOnlyTestInt;
+
+	int32 PreNetNumTimesCalled;
```

Comment:
Thinking about it some more I think a much cleaner solution is to just track `TArray<RepStep>` push on to that each call and validate that against a desired `TArray`. The unconnected integers here are still quite brittle and I think using `TArray` it will be much easier to reason about the code.
```
TArray<RepStep> RepSteps;
TArray<RepStep> ExpectedRepSteps = {RepStep::PreNetReceive, RepStep::PostNetReceive,  RepStep::RepNotify, RepStep::PreNetReceive}

void USpatialTestNetReceiveSubobject::PostNetReceive()
{
	RepSteps.Push(RepStep::PostNetReceive);
}

const SIZE_T steps = FMath::Max(Repstep.Num(), ExpectedRepSteps.Num());
for ( SIZE_T i = 0; i < steps; ++i)
{
	Step = Repstep.IsValidIndex(i) ? Repstep[i] : RepStep::None;
	ExpectedStep = ExpectedRepSteps.IsValidIndex(i) ? ExpectedRepSteps[i] : RepStep::None;
	AssertTrue(Step == ExpectedStep, FString::Printf(TEXT("Got RepStep:%s expected RepStep:%s), UEnum::GetValueAsString(Step), UEnum::GetValueAsString(ExpectedStep))
}
.. etc

```

Reply:
you're right, that is a much cleaner approach - have made the changes